### PR TITLE
Initial version for in memory system.

### PR DESCRIPTION
**Pull request by bharathkk**
Getting the PR out to unblock sanil.
Pending tasks
 1. Add documentation
 2. Add more tests
 3. Currently initialization of the stream happens using createStream on admin. We need to find a way to expose the same functionality to low level users.
 4. To populate the initial stream, clients need to get a handle on producer and produce the messages. Alternatively, we can support serialization of source data and pass it as config to the system to initialize. We have a hook in place for this but not implemented it completely.
  5. Clean up consumed data on the buffer based on the lowest offsets of the consumers.

I will create JIRAs for all the pending tasks and fix them iteratively.

@xinyuiscool ^^

### CODE REVIEW

**Comment by Sanil15** [comment]
Thanks a lot for this Bharat, 

In regards to the pending task points above mentioned, here is my update:
3. This can be avoided by just spinning up a producer and producing to input stream, I will have a prototype for low level (Async and Sync) out soon, you can refer to that 
4. This too can be avoided since spinning up a producer in start initializes the stream, I have a working prototype for it. The idea of serializing it and again reading from config is redundant

One more thing to be considered in the system is the support of bounded streams, i.e we need to put an EndOfStreamMessage in a stream partition, this can only be done in consumer poll() as of current design.

Thanks

File: samza-core/src/main/java/org/apache/samza/system/inmemory/InMemoryManager.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+  private List<IncomingMessageEnvelope> poll(SystemStreamPartition ssp, String offset) {
+    int startingOffset = Integer.parseInt(offset);
+    List<IncomingMessageEnvelope> messageEnvelopesForSSP = bufferedMessages.getOrDefault(ssp, new LinkedList<>());
+
+    if (startingOffset >= messageEnvelopesForSSP.size()) {
+      return new ArrayList<>();
+    }
+
+    return messageEnvelopesForSSP.subList(startingOffset, messageEnvelopesForSSP.size());
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/system/inmemory/InMemoryManager.java]: We do not remove the elements polled from the ssp?

File: samza-core/src/main/java/org/apache/samza/system/inmemory/InMemorySystemAdmin.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+   * offset1 == offset2 and offset1 &gt; offset2 respectively. Return
+   * null if those two offsets are not comparable
+   *
+   * @param offset1 First offset for comparison.
+   * @param offset2 Second offset for comparison.
+   * @return -1 if offset1 &lt; offset2; 0 if offset1 == offset2; 1 if offset1 &gt; offset2. Null if not comparable
+   */
+  @Override
+  public Integer offsetComparator(String offset1, String offset2) {
+    int o1 = Optional.ofNullable(offset1)
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/system/inmemory/InMemorySystemAdmin.java]: this logic is too hard to understand. I would refer a simple something like:

if(offset1 == null || offset2 == null) {
   return null;
} else {
   return Integer.valueOf(offset1).compare(Interger.valueOf(offset2));
}

File: samza-core/src/main/java/org/apache/samza/system/inmemory/InMemorySystemAdmin.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+   * Create a stream described by the spec.
+   *
+   * @param streamSpec  The spec, or blueprint from which the physical stream will be created on the system.
+   * @return            {@code true} if the stream was actually created and not pre-existing.
+   *                    {@code false} if the stream was pre-existing.
+   *                    A RuntimeException will be thrown if creation fails.
+   */
+  @Override
+  public boolean createStream(StreamSpec streamSpec) {
+    return Optional.ofNullable(inMemorySystemConfig.getSerializedDataSet(streamSpec.getId()))
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/system/inmemory/InMemorySystemAdmin.java]: I think you should just do a inMemoryManager.intializeStream(streamSpec, serializedData). Do a null check inside the function.

File: samza-core/src/main/java/org/apache/samza/system/inmemory/InMemoryManager.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  /**
+   * Populate the metadata for the {@link SystemStream} and initialize the buffer for {@link SystemStreamPartition}.
+   *
+   * @param streamSpec stream spec for the stream to be initialized
+   *
+   * @return true if successful, false otherwise
+   */
+  public boolean initializeStream(StreamSpec streamSpec) {
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/system/inmemory/InMemoryManager.java]: No need for this function. Use the following function with a null check.

File: samza-core/src/main/java/org/apache/samza/system/inmemory/InMemoryManager.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  public InMemoryManager() {
+    bufferedMessages = new ConcurrentHashMap<>();
+    systemStreamToPartitions = new ConcurrentHashMap<>();
+  }
+
+  private List<IncomingMessageEnvelope> newSynchronizedLinkedList() {
+    return  Collections.synchronizedList(new LinkedList<IncomingMessageEnvelope>());
+  }
+
+  public void put(SystemStreamPartition ssp, Object message) {
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/system/inmemory/InMemoryManager.java]: no need for this function. Use the following one with a null check or something.

File: samza-core/src/main/java/org/apache/samza/system/inmemory/InMemoryManager.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+import org.apache.samza.system.SystemStreamMetadata;
+import org.apache.samza.system.SystemStreamPartition;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Initial draft of in-memory manager. It is test only and not meant for production use right now.
+ */
+public class InMemoryManager {
```
Comment (xinyuiscool) [inline samza-core/src/main/java/org/apache/samza/system/inmemory/InMemoryManager.java]: This class should be package private and all the methods inside shouldn't be visible to public.

**Comment by prateekm** [comment]
+1 on Xinyu's point about Java8 streams API making things harder to read. Would also prefer standard constructs (if-else/ for-each).

**Pull request closed without merging**